ujt/callbacks/change_over_time_callbacks.py

```diff
@@ -38,20 +50,183 @@ def update_time_select_panel(tag):
+                seconds=temp_datetime.second,
+            )
+
+            tag_timestamp_string = tag_selection.split("@")[-1]
+            tag_timestamp = dt.datetime.strptime(
+                tag_timestamp_string, constants.DATE_FORMAT_STRING
+            )
+            start_time = tag_timestamp - window_size_timedelta
+            end_time = tag_timestamp + window_size_timedelta
+    except (ValueError, TypeError):
+        # ValueError occurs when input format is incorrect
+        # TypeError occurs when input box is blank
+        return dash.no_update, True, "Error parsing time input, please check format."
+
+    sli_response = rpc_client.get_slis(
+        start_time=start_time,
+        end_time=end_time,
+        sli_types=[sli_type],
+    )
+
+    dict_slis: List[Dict[str, Any]] = [
+        json_format.MessageToDict(sli) for sli in sli_response.slis
+    ]
+
+    # Dash serializes the objects placed into Stores via json, but this can lead to
+    # unexpected behavior with objects (particularly datetimes).
+    # This can make it confusing to store objects directly, so we choose to store
+    # timestamps instead of datetimes for a clearer interface from the reader's perspective.
+    # We are forced to convert SLI protos to dictionaries to store them as well.
+    return [
+        {
+            "start_timestamp": start_time.timestamp(),
+            "end_timestamp": end_time.timestamp(),
+            "dict_slis": dict_slis,
+        },
+        False,
+        None,
+    ]
+
+
+@app.callback(
+    Output(id_constants.CHANGE_OVER_TIME_TEXT_OUTPUT_PANEL, "children"),
+    Input(id_constants.CHANGE_OVER_TIME_SLI_STORE, "data"),
+    prevent_initial_call=True,
+)
+def update_change_over_time_text_output_panel(change_over_time_data):
+    """Renders the datatable that displays SLI changes over time.
+
+    Args:
+        change_over_time_data: Either an empty dictionary, or a dictionary with keys
+            "start_timestamp" mapped to a float POSIX timestamp,
+            "end_timestamp" mapped to a float POSIX timestamp, and
+            "dict_slis" mapped to a list of SLIs represented as dictionaries.
+            The SLIs as dictionaries need to be parsed by the json_format module.
+
+    Returns:
+        A component to be placed in the CHANGE_OVER_TIME_TEXT_OUTPUT_PANEL
+    """
+    if change_over_time_data == {}:
+        return None
+
+    start_time = dt.datetime.fromtimestamp(change_over_time_data["start_timestamp"])
+    end_time = dt.datetime.fromtimestamp(change_over_time_data["end_timestamp"])
+    dict_slis = change_over_time_data["dict_slis"]
+    slis = [json_format.ParseDict(dict_sli, SLI()) for dict_sli in dict_slis]
+
+    node_name_sli_map = defaultdict(list)
+    for sli in slis:
+        node_name_sli_map[sli.node_name].append(sli)
+
+    composite_slis = [
+        transformers.generate_before_after_composite_slis(slis, start_time, end_time)
+        for slis in node_name_sli_map.values()
+    ]
```

Comment:
Unfortunately, this code is repeated from [transformers.apply_change_over_time_classes](https://github.com/googleinterns/userjourneytool/pull/61/files#diff-b8dd0fa1b63ac7a9b8efd618a78a47b821cdf8e65f2b93c80507d80133b9f47cR98) and [graph_callbacks.update_graph_elements](https://github.com/googleinterns/userjourneytool/pull/61/files#diff-a3ce93b60d762ff953c530ff5dd922394db91a029ca14283721de8a4716b7064R212)

This contributes to the performance degradation. We could avoid this code duplication by computing and saving the list of composite SLIs in update_change_over_time_sli_store, but I feel that this makes the API too restrictive. I can imagine other instances where we would want the raw data (not in aggregate). Although not ideal, I think it's better to delegate the processing responsibility to the readers (callers) of the store.

Reply:
Why not store both versions? If we're not currently doing anything with the raw data but we want to make it available, we can just stick it in the data store. On a user-facing system without strong memory constraints we should focus on latency over saving storage.

Reply:
That makes sense, thank you!